tensorflow/java/src/main/java/org/tensorflow/Tensor.java

```diff
@@ -237,6 +591,50 @@ long getNativeHandle() {
 
   private Tensor() {}
 
+  // buffer support
+
+  private ByteBuffer buffer() {
+    return buffer(nativeHandle).order(ByteOrder.nativeOrder());
+  }
+
+  private static IllegalArgumentException incompatibleBuffer(DataType dataType) {
+    return new IllegalArgumentException("incompatible buffer for Tensor of DataType " + dataType);
+  }
+
+  private static IllegalArgumentException incompatibleBuffer(long[] shape) {
+    return new IllegalArgumentException("incompatible buffer for Tensor of shape " + Arrays.toString(shape));
+  }
+
+  private static int numElements(long[] shape) {
+    // assumes a fully-known shape
+    int n = 1;
+    for(int i = 0; i < shape.length; i++) {
+      n *= shape[i];
+    }
+    return n;
+  }
+
+  private static int elemByteSize(DataType dataType) {
+    switch(dataType) {
+      case FLOAT:
+        return Float.SIZE / Byte.SIZE;
```

Comment:
It may just be clearer to return constants here. The byte sizes for TensorFlow data types are fixed, so let's return those?

```java
case FLOAT:
case INT32:
   return 4;
case DOUBLE:
case INT64:
  return 8;
case BOOL:
  return 1;
```